### PR TITLE
feat(storage): maintain local related sst index

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Maintain the local related sst index as a pruned `HummockVersion` through applying hummock version delta.

For the version delta created from a `commit_epoch`, we will only add the sst in the `sync_uncommitted_data`. For the version delta created from a compaction task, we will only apply the inserted sst when any sst is deleted in the pruned `HummockVersion`.

Some previous code is modified accordingly. In previous code, no matter whether the version delta is generated from a `commit_epoch` or a compaction task, in `apply_version_delta`, it will call `apply_compact_ssts`. The naming is confusing, and the implementation mixes the two cases together. In this PR, we will separate the two cases by whether the max committed epoch increases. If the max committed epoch increases, it will be generated by `commit_epoch`, and there will by only SST insertion, and a new L0 sub level will be created, and otherwise, it will be generated by a compaction, and call the `apply_compact_ssts`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

#4072

### CODE REVIEW

**Comment by Little-Wallace** [comment]
How about create a new structure just for state-store read to replace `HummockVersion`?
I think a no-proto structure would be easier to use and we can change the component such as filter some SSTs.

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs [hunk 1]
```diff
@@ -162,109 +189,205 @@ impl HummockVersionExt for HummockVersion {
...
+    fn get_level_mut(&mut self, idx: usize) -> &mut Level;
+    fn apply_compact_ssts(&mut self, summary: LevelDeltasSummary, local_related_only: bool);
+}
+
+impl HummockLevelsExt for Levels {
+    fn get_level0(&self) -> &OverlappingLevel {
+        self.l0.as_ref().unwrap()
+    }
+
+    fn get_level(&self, level_idx: usize) -> &Level {
```
Comment (soundOfDestiny) [inline src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:252]: guess we'd better not minus here avoid confusion

**Review by soundOfDestiny — COMMENTED** [review]
May you describe your new sub-level design in this PR? thanks

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs (hunk 1 above)
Comment (wenym1) [reply src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:252]: The code is simply moved from L305 in the current main branch in the same file. We'd better keep it consistent with previous code.

**Comment by wenym1** [comment]
> May you describe your new sub-level design in description of this PR? thanks

The current sub-level design is untouched. I did some simple refactoring so that we can reuse most of current code in this PR. 

The most modified refactoring will be separating the apply logic of version delta generated from `commit_epoch` and a compaction task. I have added description about this refactoring in the PR description.

**Review by hzxa21 — COMMENTED** [review]
Rest LGTM.

File: src/storage/src/hummock/local_version.rs [hunk 2]
```diff
@@ -337,28 +372,180 @@ impl LocalVersion {
...
+        &mut self,
+        max_committed_epoch: HummockEpoch,
+    ) -> Vec<(Vec<HummockEpoch>, Vec<LocalSstableInfo>)> {
+        self
+            .sync_uncommitted_data
+            .drain_filter(|(epoch, data)| {
+                let min_epoch = *epoch.first().expect("epoch list should not be empty");
+                let max_epoch = *epoch.last().expect("epoch list should not be empty");
+                assert!(
+                    max_epoch <= max_committed_epoch
```
[the remark this answers is not in the document]
Comment (wenym1) [reply src/storage/src/hummock/local_version.rs:386]: Here we want to ensure that either all epochs are synced or no epoch is synced. 

Given a max_committed_epoch, we will sync any epoch with `epoch<=max_committed_epoch`. For `max_epoch<=max_committed_epoch`, it means all epochs are synced. For `min_epoch >max_committed_epoch`, it means no epoch is synced.

**Comment by soundOfDestiny** [comment]
guess we can merge it now

**Review by hzxa21 — APPROVED** [review]
LGTM

**Pull request merged**